### ISSUE TITLE
Allow non time based history merging

### ISSUE THREAD

**Issue by georeith** [yjs/yjs#290]
**Checklist**

[ ] Are you reporting a bug? Use github issues for bug reports and feature requests. For general questions, please use https://discuss.yjs.dev/
[ ] Try to report your issue in the correct repository. Yjs consists of many modules. When in doubt, report it to https://github.com/yjs/yjs/issues/

**Is your feature request related to a problem? Please describe.**

Outside of text editing there are many operations that may be long lived, e.g., drag operations, where the user may continue to move an object, pause and then move again without releasing the object. Typically you would want this operation to be a single event in history.

**Describe the solution you'd like**

It would be great if alongside tracked origins we could specify an additional `transactionName` and if the following transaction's name matches the preceding transaction's name they will become a single event in history regardless of the time delta between the two transactions.

**Describe alternatives you've considered**
Could manage this in userland by storing the `groupName` in metadata and manually undoing/redoing while `groupName` matches too.

**Comment by dmonad** [comment]
Hi @georeith, that makes sense. It might already be possible to do this by increasing the timeout to infinity (there is no public API for that yet though), and then always calling `stopCapturing()` manually. 

In your suggested approach, would `transactionName` be separate from `origin`? Would we need to adapt `doc.transact(..)` just for this use-case, or is there maybe a simpler approach?

**Comment by georeith** [comment]
Hey @dmonad thanks for the swift reply, this suggestion came ahead of me porting my an existing state management to Yjs where we had an undo manager that did both this and time based grouping of events.

The time based is useful as a baseline as it catches most things and it can be argued that 2 interactions happening within a certain period would never want to be considered as anything but a single event in history by a user as they couldn't be doing two distinct things that quickly.

If this was added as a feature, for my use case I wouldn't want it in lieu of time based but as an addition to so that you could use both at the same time (an event groups if the name matches or if the delta is within a certain limit).

In terms of the API around it I am not too sure, I was thinking an additional parameter to `doc.transact()` (perhaps a metadata object that could be extended in future), there may be a simpler approach and would love to see some alternatives, the `transactionName` was just the simplest way we found to do it originally (both to handle at the event merging point and for usage in the tool). The `transactionName` would typically end up looking something like `resize-1617875006556` with the epoch timestamp being when the initial mousedown occured for that resize action.

> In your suggested approach, would transactionName be separate from origin?

Ideally yes, as in you would never merge two events from a different origin regardless of whether their `transactionName` matched.

**Comment by dmonad** [comment]
I've been thinking for quite some time to add something like a `metadata` object to the transaction for assigning custom information to a transaction. We could reuse that to allow your use-case. 

Personally, I like the approach to name/describe transactions and then merge UndoManager stack-items based on their event-type. However, I think the suggested solution is too specific. Rather, we should improve Y.UndoManager to allow custom behavior. Once the metadata feature on transactions is implemented, you should be able to configure the UndoManager in a way to implement your solution.

**Comment by georeith** [comment]
@dmonad love the custom matching idea.

**Comment by dmonad** [comment]
Yeah, that makes sense. Could you take a stab at it? You only need to add another condition to the current timeout implementation.

**Comment by georeith** [comment]
@dmonad Yeah I will take a stab at it when I get to that point in moving our current implementation over 👍 .
